packages/edit-post/src/hooks/components/index.js

```diff
@@ -11,3 +18,78 @@ addFilter(
 	'core/edit-post/replace-media-upload',
 	replaceMediaUpload
 );
+
+function getMediaSourceUrlBySizeSlug( media, slug ) {
+	return (
+		media?.media_details?.sizes?.[ slug ]?.source_url || media?.source_url
+	);
+}
+
+const addFeaturedImageToolbarItem = createHigherOrderComponent(
+	( BlockEdit ) => {
+		return ( props ) => {
+			const {
+				attributesContext = [],
+				attributes,
+				setAttributes,
+				name: blockName,
+			} = props;
+
+			const { useFeaturedImage } = attributes;
+
+			const { sizeSlug } = attributes;
+			const [ featuredImage ] = useEntityProp(
+				'postType',
+				'post',
+				'featured_media'
+			);
+
+			const media = useSelect(
+				( select ) =>
+					featuredImage &&
+					select( coreStore ).getMedia( featuredImage, {
+						context: 'view',
+					} ),
+				[ featuredImage ]
+			);
+			const mediaUrl = getMediaSourceUrlBySizeSlug( media, sizeSlug );
+
+			if ( useFeaturedImage ) {
+				attributesContext.push( { url: mediaUrl } );
+			}
+
+			if (
+				( 'core/image' !== blockName && 'core/cover' !== blockName ) ||
+				! mediaUrl
+			) {
+				return <BlockEdit { ...props } />;
+			}
+			return (
+				<>
+					<BlockControls group="other">
+						<ToolbarButton
+							icon={ group /*this is temporary*/ }
+							label={ __( 'Use featured image' ) }
+							onClick={ () => {
+								setAttributes( {
+									useFeaturedImage: ! useFeaturedImage,
+								} );
+							} }
+						/>
+					</BlockControls>
+					<BlockEdit
+						{ ...props }
+						attributesContext={ attributesContext }
```

Comment:
Would some variation of `{ ...Object.assign( {}, ...attributesContext ) }` also do the trick?

Reply:
Why? I find this a lot more readable, if we move forward with this direction (instead of your suggested approach to hook into data) this `attributesContext` is a special prop.

For public record I am not a fan of saving characters in interpreted programming 😁

Reply:
Oh sorry, I wasn’t after saving a few characters. I hoped we could spread the properties in here instead of passing the attributesContext at all.

Reply:
Great. I do think that in the end this has to move away from the approach here, and more into managing block data directly, like you suggested.